### PR TITLE
plugin and parameter enhancements

### PR DESCRIPTION
**Pull request by rustyrussell**
~Still a WIP, since it's based on a merge of master and #2141 which I both needed.~

Goodies inside:
* Plugin infrastructure cleanups (neatenings, pass-through)
* More plugin debugability: --dev-debugger support, stderr preservation and printing bad JSON.
* Large reorganization of param.c so it can be moved to common so plugins can share.

@wythe please take a look too, since some disentangling was required!
(BTW, please accept https://github.com/ElementsProject/lightning/invitations so I can tag you for review and bug you in future :)

### CODE REVIEW

File: lightningd/jsonrpc.c [hunk 1]
```diff
@@ -981,14 +976,20 @@ static bool json_tok_command(struct command *cmd, const char *name,
 	return false;
 }
 
+/* We add this destructor as a canary to detect cmd failing. */
+static void destroy_command_canary(struct command *cmd, bool *failed)
+{
+	*failed = true;
+}
+
```
Comment (wythe) [inline lightningd/jsonrpc.c]: I love it!  So  much better.

**Comment by rustyrussell** [comment]
Trivial rebase on top of master now requirements have been merged...

File: common/param.h [hunk 2]
```diff
@@ -90,4 +94,6 @@ typedef bool(*param_cbx)(struct command *cmd,
 				   (const jsmntok_t *)NULL,	    \
 				   (arg)) == true); })
 
-#endif /* LIGHTNING_LIGHTNINGD_PARAM_H */
+/* Special flag for 'check' which allows any parameters. */
+#define p_opt_any() "", false, NULL, NULL
```
Comment (wythe) [inline common/param.h]: Nicely done. This approach crossed my mind at some point but then it was gone. Its clearly better 👍

**Comment by wythe** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2148/commits/c31e59d178dba6af66b1c663bb387157f0c2d36d

param/json stuff looks great.

**Pull request merged**